File: app/src/main/java/vn/loitp/app/activity/picker/timepicker/RangeTimePickerDialog.java
package vn.loitp.app.activity.picker.timepicker;

import android.app.TimePickerDialog;
import android.content.Context;
import android.widget.TimePicker;

import java.lang.reflect.Field;

public class RangeTimePickerDialog extends TimePickerDialog {
    private int mMinHour = -1;
    private int mMinMinute = -1;
    private int mMaxHour = 100;
    private int mMaxMinute = 100;
    private int mCurrentHour;
    private int mCurrentMinute;

    public RangeTimePickerDialog(Context context, OnTimeSetListener callBack, int hourOfDay,
                                 int minute, boolean is24HourView) {
        super(context, callBack, hourOfDay, minute, is24HourView);
        mCurrentHour = hourOfDay;
        mCurrentMinute = minute;
        // Somehow the onTimeChangedListener is not set by TimePickerDialog
        // in some Android Versions, so, Adding the listener using
        // reflections
        try {
            Class<?> superclass = getClass().getSuperclass();
            if (superclass != null) {
                Field mTimePickerField = superclass.getDeclaredField("mTimePicker");
                mTimePickerField.setAccessible(true);
                TimePicker mTimePicker = (TimePicker) mTimePickerField.get(this);
                if (mTimePicker != null) {
                    mTimePicker.setOnTimeChangedListener(this);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setMin(int hour, int minute) {
        mMinHour = hour;
        mMinMinute = minute;
    }

    public void setMax(int hour, int minute) {
        mMaxHour = hour;
        mMaxMinute = minute;
    }

    @Override
    public void onTimeChanged(TimePicker view, int hourOfDay, int minute) {
        super.onTimeChanged(view, hourOfDay, minute);
        boolean validTime = ((hourOfDay >= mMinHour) && (hourOfDay != mMinHour || minute >= mMinMinute))
                && ((hourOfDay <= mMaxHour) && (hourOfDay != mMaxHour || minute <= mMaxMinute));
        if (validTime) {
            mCurrentHour = hourOfDay;
            mCurrentMinute = minute;
        } else {
            updateTime(mCurrentHour, mCurrentMinute);
        }
    }
}

File: app/src/main/java/vn/loitp/up/a/picker/time/RangeTimePickerDialog.kt
package vn.loitp.up.a.picker.time

import android.app.TimePickerDialog
import android.content.Context
import android.widget.TimePicker

class RangeTimePickerDialog(
    context: Context,
    callBack: OnTimeSetListener? = null,
    private var mCurrentHour: Int,
    private var mCurrentMinute: Int,
    is24HourView: Boolean = true
) : TimePickerDialog(context, callBack, mCurrentHour, mCurrentMinute, is24HourView) {

    private var mMinHour = -1
    private var mMinMinute = -1
    private var mMaxHour = 100
    private var mMaxMinute = 100

    fun setMin(
        hour: Int,
        minute: Int
    ) {
        mMinHour = hour
        mMinMinute = minute
    }

    fun setMax(
        hour: Int,
        minute: Int
    ) {
        mMaxHour = hour
        mMaxMinute = minute
    }

    override fun onTimeChanged(
        view: TimePicker,
        hourOfDay: Int,
        minute: Int
    ) {
        super.onTimeChanged(view, hourOfDay, minute)

        val validTime = (
                hourOfDay >= mMinHour &&
                        (hourOfDay != mMinHour || minute >= mMinMinute) &&
                        hourOfDay <= mMaxHour &&
                        (hourOfDay != mMaxHour || minute <= mMaxMinute)
                )
        if (validTime) {
            mCurrentHour = hourOfDay
            mCurrentMinute = minute
        } else {
            updateTime(mCurrentHour, mCurrentMinute)
        }
    }

    init {
        // Somehow the onTimeChangedListener is not set by TimePickerDialog
        // in some Android Versions, so, Adding the listener using
        // reflections
        try {
            val superclass: Class<*>? = javaClass.superclass
            if (superclass != null) {
                val mTimePickerField = superclass.getDeclaredField("mTimePicker")
                mTimePickerField.isAccessible = true
                val mTimePicker = mTimePickerField[this] as TimePicker
                mTimePicker.setOnTimeChangedListener(this)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
